flink-runtime/src/main/java/org/apache/flink/runtime/client/JobClient.java

```diff
@@ -160,12 +159,12 @@ public static SerializedJobExecutionResult submitJobAndWait(
 
 				SerializedJobExecutionResult result = ((JobManagerMessages.JobResultSuccess) answer).result();
 				if (result != null) {
-					return result;
+					return result.toJobExecutionResult(usercodeClassloader);
 				} else {
 					throw new Exception("Job was successfully executed but result contained a null JobExecutionResult.");
 				}
-			} else if (answer instanceof Status.Failure) {
-				throw ((Status.Failure) answer).cause();
+			} else if (answer instanceof JobClientMessages.JobFailure) {
```

Comment:
Why do we introduce a new `JobFailure` message here? Can't we simply let our `SerializedThrowable` class inherit from `Exception` and send it as a failure over the network? We then only have to add in the outer try block a catch block which deserializes the wrapped exception.